plugins/pulsar-records-storage/src/main/java/com/github/bsideup/liiklus/pulsar/PulsarRecordsStorage.java

```diff
@@ -0,0 +1,196 @@
+                                    .hashingScheme(HashingScheme.Murmur3_32Hash)
+                                    .enableBatching(false)
+                                    .createAsync()
+                    ).cache();
+                })
+                .flatMap(producer -> {
+                    val valueBytes = new byte[envelope.getValue().remaining()];
+                    envelope.getValue().duplicate().get(valueBytes);
+                    val future = producer.newMessage()
+                            .key(StandardCharsets.UTF_8.decode(envelope.getKey().duplicate()).toString())
+                            .value(valueBytes)
+                            .sendAsync();
+
+                    return Mono.fromCompletionStage(future)
+                            .cast(MessageIdImpl.class)
+                            .map(it -> new OffsetInfo(
+                                    topic,
+                                    it.getPartitionIndex(),
+                                    toOffset(it)
+                            ));
+                })
+                .toFuture();
+    }
+
+    @Override
+    public Subscription subscribe(String topic, String groupName, Optional<String> autoOffsetReset) {
+        return new PulsarSubscription(topic, groupName, autoOffsetReset);
+    }
+
+    @Value
+    private class PulsarSubscription implements Subscription {
+
+        String topic;
+
+        String groupName;
+
+        Optional<String> autoOffsetReset;
+
+        @Override
+        public Publisher<Stream<? extends PartitionSource>> getPublisher(
+                Supplier<CompletionStage<Map<Integer, Long>>> offsetsProvider
+        ) {
+            return Mono
+                    .defer(() -> Mono.fromCompletionStage(((PulsarClientImpl) pulsarClient).getNumberOfPartitions(topic)))
+                    .mergeWith(Flux.never()) // Never complete
+                    .map(numberOfPartitions -> {
+                        return IntStream.range(0, numberOfPartitions).mapToObj(partition -> new PulsarPartitionSource(
+                                topic,
+                                partition,
+                                groupName,
+                                autoOffsetReset,
+                                Mono.defer(() -> Mono
+                                        .fromCompletionStage(offsetsProvider.get())
+                                        .handle((it, sink) -> {
+                                            if (it.containsKey(partition)) {
+                                                sink.next(it.get(partition));
+                                            } else {
+                                                sink.complete();
+                                            }
+                                        })
+                                )
+                        ));
+                    });
+        }
+
+        @Override
+        public boolean equals(Object o) {
```

Comment:
should be hashcode implemented too?

Reply:
Works like that, only implemented because of the tests